### PR TITLE
Implement 2FA reset for T3

### PR DESCRIPTION
**Pull request by tomlinton**
Implements 2FA reset for T3.

### Checklist:

- [ ] Test your work and double-check to confirm that you didn't break anything
- [ ] [Wrap any new text/strings with `fbt`](https://github.com/OriginProtocol/origin/tree/master/dapps/marketplace/translation#wrapping-text) for translation
- [ ] If there are any new/changed strings to translate, [`npm run translate`](https://github.com/OriginProtocol/origin/tree/master/dapps/marketplace/translation#integrating-translations)
- [ ] Update any relevant READMEs and [docs](https://github.com/OriginProtocol/origin/tree/master/docs)

### CODE REVIEW

File: infra/token-transfer-client/src/actions/otp.js [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+function submitOtpUpdatePending() {
+  return {
+    type: SUBMIT_OTP_UPDATE_PENDING
+  }
+}
+
+function submitOtpUpdateSuccess(payload) {
+  return {
+    type: SUBMIT_OTP_UPDATE_SUCCESS,
+    payload
```
Comment (shahthepro) [inline infra/token-transfer-client/src/actions/otp.js:40]: nit: `payload` param doesn't seem to be passed anywhere in the reducer for this action. Could remove it.

File: infra/token-transfer-client/src/components/OtpModal.js [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+    this.state = this.getInitialState()
+  }
+
+  getInitialState = () => {
+    const initialState = {
+      oldCode: '',
+      oldCodeError: null,
+      newCode: '',
+      newCodeError: null,
+      modalState: 'OTP'
```
Comment (shahthepro) [inline infra/token-transfer-client/src/components/OtpModal.js:32]: nit: Probably better to put this as an enum

**Review by shahthepro — APPROVED** [review]
Had a quick look, looks good to me. LGTM

File: infra/token-transfer-client/src/actions/otp.js (hunk 1 above)
Comment (tomlinton) [reply infra/token-transfer-client/src/actions/otp.js]: 👍

**Review by franckc — APPROVED** [review]
LGTM

File: infra/token-transfer-server/src/controllers/user.js [hunk 3]
```diff
@@ -95,6 +106,54 @@ router.post(
...
+    if (!errors.isEmpty()) {
+      return res
+        .status(422)
+        .json({ errors: errors.array({ onlyFirstError: true }) })
+    }
+
+    if (req.body.newCode) {
+      // If newCode is in body we are updating the uer model with the new
+      // OTP key
+      const encryptedKey = encrypt(req.body.otpKey)
```
Comment (franckc) [inline infra/token-transfer-server/src/controllers/user.js:130]: Should we add a defensice check to enforce req.body.otpKey is not empty?

**Comment by franckc** [comment]
Do we have any sort of alerting on T3? It would be nice to have an alert on unusual high volume of 2FA resets... that could indicate a hack.

**Comment by tomlinton** [comment]
We've got Discord webhooks so I'll add a notification 👍

**Comment by franckc** [comment]
> We've got Discord webhooks so I'll add a notification 👍

ah yes, perfect!

**Pull request merged**